### ISSUE TITLE
Spread operator is shown as an issue even when using Kotlin 1.6.21

### ISSUE THREAD

**Issue by Abhimanyu14** [detekt/detekt#4860]
As per this Docs - https://detekt.dev/docs/rules/performance#spreadoperator,  
If we use Kotlin above version 1.1.60, the spread operator should not be a performance issue,

## Expected Behavior
When using the spread operator in a project with Kotlin 1.6.21, it should not be reported as an issue.

## Observed Behavior
The spread operator is reported as an issue

## Your Environment

* Version of detekt used: 1.18.0
* Version of Gradle used (if applicable):
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version:
* Link to your project (if it's a public repository): https://github.com/Abhimanyu14/finance-manager

**Comment by cortinico** [comment]
> When type resolution is enabled in detekt this case will not be flagged by the rule since it doesn't suffer the performance penalty of an array copy.

Are you using type resolution?

**Comment by Abhimanyu14** [comment]
Tried with `./gradlew detektDebug` as mentioned [here](https://detekt.dev/docs/gettingstarted/type-resolution/#enabling-on-an-android-project)

Now, this is the error - SpreadOperator - [Used in this way a spread operator causes a full copy of the array to be created(…)] at <file_path>

The spread is used for varargs.

**Comment by BraisGabin** [comment]
This is not the first issue raisede because of this rule. I know that the first versions of Kotlin had this performance issue. But is this still an issue in new versions of Kotlin? If it is, when that performance impact happen exactly?

If there is proper documentation about this performance issue I think that we can fix it. But if there is not I would vote to remove it.

**Comment by cortinico** [comment]
I've briefly looked into this,

> The spread is used for varargs.

If this is the case, then this looks like a bug in the rule. Please provide a snippet of the reproducer. In general, you should not be flagged on usages of spread operator on varargs.

> But is this still an issue in new versions of Kotlin? If it is, when that performance impact happen exactly?

Any further reference would be helpful to know what's the current status. We might just deprecate the rule entirely if that's the case.

**Comment by Abhimanyu14** [comment]
This is the code

```
insertTransactionsUseCase(
  *transactions.toTypedArray(),
)
```

The type signature of `transactions`
```
val transactions: List<Transaction>
```

And the Detekt issue

> SpreadOperator - [Used in this way a spread operator causes a full copy of the array to be created(…)] at /Users/abhimanyu/Documents/projects/finance-manager/codebase/android/app/src/main/java/com/makeappssimple/abhimanyu/financemanager/android/data/usecase/RestoreDataUseCase.kt:56:34

Kindly check this public repo for more details - https://github.com/Abhimanyu14/finance-manager

**Comment by cortinico** [comment]
Thanks for the repro. Pinging @3flex as he originally wrote the rule. I'm unsure if your case is a false positive or not.

A test case for your scenario is:

```kotlin
        @Test
        fun `test case for #4860`() {
            val code = """
            fun test0(vararg strs: String) {
                // do nothing
            }

            fun test(params : List<String>) {
                val array = params.toTypedArray()
                test0(*array)
            }
            """
            assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
        }
```

if someone wants to pick this up 👍

**Comment by gouri-panda** [comment]
@cortinico Can i take this?

**Comment by 3flex** [comment]
The specific optimisation introduced in Kotlin 1.1.60 was to avoid an array copy specifically when using `*<array-constructor-fun>(...)`: https://youtrack.jetbrains.com/issue/KT-20462

It was quite specific in the optimisation that it introduced. It didn't avoid the array copy for every single use of the spread operator.

If there were further optimisations in the compiler since 1.1.60 I'm not aware of them, but if there have been any, then we can update the rule accordingly. If there haven't been any changes since 1.1.60 then the current rule should still be reporting this issue correctly, including https://github.com/detekt/detekt/issues/4860#issuecomment-1145557237.

**Comment by cortinico** [comment]
Thanks for the clarification @3flex. I'm not aware of any other optimization on the spread operator. Maybe @gouri-panda or someone else wants to investigate further?

Otherwise we can close this as "expected behavior"

**Comment by gouri-panda** [comment]
@cortinico Sure! I can investigate further :)

**Comment by 3flex** [comment]
I'm going to close this for now as there's no current action to take.

If there are any more recent compiler optimisations identified that need to be taken into account please either create a new issue, or respond to this issue and we can reopen, thanks!

**Comment by eygraber** [comment]
I don't want to read through the whole thing so I'm not fully sure, but does https://github.com/JetBrains/kotlin/pull/2510 optimize any of the cases that detekt currently warns about?

**Comment by MukjepScarlet** [comment]
If I want to use a Java function like `CompletableFuture.allOf`, what can I do for this issue?
